### DISCUSSION TITLE
Retrieving x, y, z Affine Transform Parameters After Coregistration

### DISCUSSION THREAD

**Discussion by cwywilson** [GlacioHack/xdem#460] [General]
Happy New Year,

Hello everyone,

I’ve been using the number of xdem.coreg to perform coregistration between two DEMs. The coregistration process works well, but I’m interested in understanding the transformation that was applied.

Specifically, I’m looking for a way to retrieve the x, y, z affine transform parameters that were used to align the two DEMs

Is there a way to get these transformation parameters after the coregistration process? Any guidance would be greatly appreciated.

Thank you!

Wilson

**Comment by rhugonnet** [comment]
Hi @cwywilson,

You can access the transform and other parameters of coreg fits in the attribute `._meta` of a `Coreg` object:

```python
nk = coreg.NuthKaab()
nk.fit(ref_dem, tba_dem, ...)
params = nk._meta
```

It is a dictionary. Typically, X/Y shifts are in pixels ("offset_..."), Z is in DEM units ("vshift").

We need to make that attribute public and document the different keys that exist for each method (and make them more consistent!), it is planned along other work to finalize the Coreg modules in the next months! :smile:

Happy new year as well!

**Reply by cwywilson** [reply]
Hi @rhugonnet 

Thanks for your help. May I borrow this opportunity to ask you for some insight?

I use the TerrainBias below to coreg two DEMs (one is photogrammetrical DEM and the other is ArcticDEM). Since the DEM is divided into bins, do you have any suggestions to present the shift in X, Y, Z for showing the coreg process?

Here is the code:

terbias = xdem.coreg.TerrainBias(terrain_attribute="elevation")
terbias.fit(reference_dem.data, dem_to_be_aligned.data, inlier_mask=inlier_mask, transform=dem_1960.transform, crs=dem_1960.crs)
corrected_dem = terbias.apply(dem_to_be_aligned)

# Retrieve the parameters
params = terbias._meta
# Extract the affine transform parameters
x_shift = params.get('offset_east_px', None)
y_shift = params.get('offset_north_px', None)
z_shift = params.get('vshift', None)

# Print the parameters
print(f"X shift: {x_shift}")
print(f"Y shift: {y_shift}")
print(f"Z shift: {z_shift}")

**Reply by rhugonnet** [reply]
One of the top priorities for the work planned on Coreg is to add better plotting/summary functions, which would help with the subject you mention. 
You're confirming the need for this. :slightly_smiling_face: 

For `TerrainBias`, there is no X/Y/Z shifts. The correction is with a function of your variable ("elevation" here), of the form chosen by you (if method is "fit" or "bin_and_fit") or only based on the binning (if method is "bin"). 

If you are using binning ("bin" or "bin_and_fit"), you can have a look at the offsets found by plotting the dataframe in `terbias._meta["bin_statistic"]`. 
If you chose to fit a functional form ("fit" or "bin_and_fit"), you can also plot your final correction function (form is stored in `terbias._meta["fit_func"]` with parameters found during optimization in `terbias._meta["fit_params"]`).

**Reply by cwywilson** [reply]
Thank you for bringing XDEM to life; it has been incredibly beneficial for my research. I'm eager to contribute more to the development of XDEM, but my coding skills are currently limited. I'm eagerly anticipating updates that might introduce additional functionalities, such as interpolation, to enhance its capabilities.
